<a href="https://colab.research.google.com/github/tohnperfect/DWDM21/blob/main/Chapter6_Association_Rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/DWDM2021_DATA'

Mounted at /content/drive


In [3]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin1')

In [4]:
data

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...    12680.0          France
541905    581587     22899  ...    12680.0          France
541906    581587     23254  ...    12680.0          France
541907    581587     23255  ...    12680.0          France
541908    581587     22138  ...    12680.0          France

[541909 rows x 8 columns]

In [6]:
data.sort_values(by=['InvoiceNo'])

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
540449   C581490     23144  ...    14397.0  United Kingdom
541541   C581499         M  ...    15498.0  United Kingdom
541715   C581568     21258  ...    15311.0  United Kingdom
541716   C581569     84978  ...    17315.0  United Kingdom
541717   C581569     20979  ...    17315.0  United Kingdom

[541909 rows x 8 columns]

In [14]:
data[data['InvoiceNo'].str.contains('C')]

InvoiceNo StockCode  ... CustomerID         Country
141      C536379         D  ...    14527.0  United Kingdom
154      C536383    35004C  ...    15311.0  United Kingdom
235      C536391     22556  ...    17548.0  United Kingdom
236      C536391     21984  ...    17548.0  United Kingdom
237      C536391     21983  ...    17548.0  United Kingdom
...          ...       ...  ...        ...             ...
540449   C581490     23144  ...    14397.0  United Kingdom
541541   C581499         M  ...    15498.0  United Kingdom
541715   C581568     21258  ...    15311.0  United Kingdom
541716   C581569     84978  ...    17315.0  United Kingdom
541717   C581569     20979  ...    17315.0  United Kingdom

[9288 rows x 8 columns]

In [17]:
data[data['InvoiceNo'].str.contains('536391')]

InvoiceNo StockCode  ... CustomerID         Country
235   C536391     22556  ...    17548.0  United Kingdom
236   C536391     21984  ...    17548.0  United Kingdom
237   C536391     21983  ...    17548.0  United Kingdom
238   C536391     21980  ...    17548.0  United Kingdom
239   C536391     21484  ...    17548.0  United Kingdom
240   C536391     22557  ...    17548.0  United Kingdom
241   C536391     22553  ...    17548.0  United Kingdom

[7 rows x 8 columns]

In [18]:
data[data['CustomerID']==17548]

InvoiceNo StockCode  ... CustomerID         Country
235      C536391     22556  ...    17548.0  United Kingdom
236      C536391     21984  ...    17548.0  United Kingdom
237      C536391     21983  ...    17548.0  United Kingdom
238      C536391     21980  ...    17548.0  United Kingdom
239      C536391     21484  ...    17548.0  United Kingdom
240      C536391     22557  ...    17548.0  United Kingdom
241      C536391     22553  ...    17548.0  United Kingdom
165024    550755     22585  ...    17548.0  United Kingdom
165025    550755     22082  ...    17548.0  United Kingdom
165026    550755     22081  ...    17548.0  United Kingdom
165027    550755     22079  ...    17548.0  United Kingdom
165028    550755     22926  ...    17548.0  United Kingdom
177223   C552049     22926  ...    17548.0  United Kingdom
177224   C552049     22585  ...    17548.0  United Kingdom
177225   C552049     22082  ...    17548.0  United Kingdom
177226   C552049     22081  ...    17548.0  United Kingdom
177227   C552049     22079  ...    17548.0  United Kingdom

[17 rows x 8 columns]

### ลบ records ที่ถูก cancel ออกไป

In [20]:
data_clean = data[data['Quantity']>=1]
data_clean

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...    12680.0          France
541905    581587     22899  ...    12680.0          France
541906    581587     23254  ...    12680.0          France
541907    581587     23255  ...    12680.0          France
541908    581587     22138  ...    12680.0          France

[531285 rows x 8 columns]

### [Q] มีประเทศสาขาของ supermarket นี้ทั้งหมดกี่ประเทศ

In [ ]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [ ]:
len(set(data['Country']))

38

### [HW13] วาดกราฟสรุปจำนวน items และ ยอดขายของแต่ละประเทศ


เพิ่ม column ราคารวม

In [22]:
data_clean['total']=data_clean['Quantity']*data_clean['UnitPrice']
data_clean

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


InvoiceNo StockCode  ...         Country  total
0         536365    85123A  ...  United Kingdom  15.30
1         536365     71053  ...  United Kingdom  20.34
2         536365    84406B  ...  United Kingdom  22.00
3         536365    84029G  ...  United Kingdom  20.34
4         536365    84029E  ...  United Kingdom  20.34
...          ...       ...  ...             ...    ...
541904    581587     22613  ...          France  10.20
541905    581587     22899  ...          France  12.60
541906    581587     23254  ...          France  16.60
541907    581587     23255  ...          France  16.60
541908    581587     22138  ...          France  14.85

[531285 rows x 9 columns]

In [23]:
data_clean.groupby(['Country']).sum()

Quantity     UnitPrice    CustomerID         total
Country                                                                 
Australia                84209  3.612500e+03  1.475853e+07  1.385213e+05
Austria                   4881  1.693900e+03  4.982601e+06  1.019868e+04
Bahrain                    314  8.276000e+01  2.100270e+05  7.541400e+02
Belgium                  23237  7.372850e+03  2.524670e+07  4.119634e+04
Brazil                     356  1.426000e+02  4.086080e+05  1.143600e+03
Canada                    2763  9.105800e+02  2.615483e+06  3.666380e+03
Channel Islands           9491  3.389650e+03  1.113834e+07  2.045044e+04
Cyprus                    6361  3.506180e+03  7.616904e+06  1.359038e+04
Czech Republic             671  7.827000e+01  3.195250e+05  8.267400e+02
Denmark                   8235  1.195550e+03  4.757518e+06  1.895534e+04
EIRE                    147447  3.848995e+04  1.067247e+08  2.834540e+05
European Community         499  2.898000e+02  9.064800e+05  1.300250e+03
Finland                  10704  3.628440e+03  8.574309e+06  2.254608e+04
France                  112104  3.699279e+04  1.057630e+08  2.097151e+05
Germany                 119263  3.353214e+04  1.143723e+08  2.288671e+05
Greece                    1557  6.632900e+02  1.995867e+06  4.760520e+03
Hong Kong                 4773  6.666740e+03  0.000000e+00  1.569180e+04
Iceland                   2458  4.812100e+02  2.247154e+06  4.310000e+03
Israel                    4409  1.070980e+03  3.139135e+06  8.135260e+03
Italy                     8112  3.576210e+03  9.588028e+06  1.748324e+04
Japan                    26016  6.572100e+02  4.095425e+06  3.741637e+04
Lebanon                    386  2.424400e+02  5.743800e+05  1.693880e+03
Lithuania                  652  9.944000e+01  5.366200e+05  1.661060e+03
Malta                      970  5.451900e+02  1.891076e+06  2.725590e+03
Netherlands             200937  6.247730e+03  3.407895e+07  2.854463e+05
Norway                   19338  5.662470e+03  1.333461e+07  3.616544e+04
Poland                    3684  1.377210e+03  4.201606e+06  7.334650e+03
Portugal                 16258  8.770720e+03  1.863478e+07  3.374710e+04
RSA                        352  2.481000e+02  7.218680e+05  1.002310e+03
Saudi Arabia                80  2.116000e+01  1.130850e+05  1.459200e+02
Singapore                 5241  1.294999e+04  2.829168e+06  2.127929e+04
Spain                    27951  9.504340e+03  3.208028e+07  6.157711e+04
Sweden                   36083  1.695790e+03  6.637579e+06  3.837833e+04
Switzerland              30630  6.634360e+03  2.332474e+07  5.708990e+04
USA                       2458  4.133000e+02  2.260641e+06  3.580390e+03
United Arab Emirates       982  2.298900e+02  1.018952e+06  1.902280e+03
United Kingdom         4733819  1.845444e+06  5.510928e+09  9.003098e+06
Unspecified               3300  1.204010e+03  3.348046e+06  4.749790e+03

In [25]:
data_clean[data_clean['InvoiceNo'].str.contains('A')]

InvoiceNo StockCode  ...         Country     total
299982   A563185         B  ...  United Kingdom  11062.06
299983   A563186         B  ...  United Kingdom -11062.06
299984   A563187         B  ...  United Kingdom -11062.06

[3 rows x 9 columns]

## เตรียม Data สำหรับ (Fequence Pattern) Association Rule

In [24]:
data_clean.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
581586                          [22061, 23275, 21217, 20685]
581587     [22631, 22556, 22555, 22728, 22727, 22726, 227...
A563185                                                  [B]
A563186                                                  [B]
A563187                                                  [B]
Name: StockCode, Length: 20728, dtype: object

In [29]:
transactions = data_clean.groupby('InvoiceNo')['StockCode'].apply(list)

## Apriori

http://zaxrosenberg.com/unofficial-apyori-documentation/

In [27]:
!pip install apyori 

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=d9d2218c7d23f8a33c8159f698abfeac24ffefb8b961b399bedb99de799c42d9
  Stored in directory: /root/.cache/pip/wheels/cb/f6/e1/57973c631d27efd1a2f375bd6a83b2a616c4021f24aab84080
Successfully built apyori


In [28]:
from apyori import apriori

In [31]:
frequentItemsets = list(apriori(transactions,min_support = 0.1))
frequentItemsets

[RelationRecord(items=frozenset({'85099B'}), support=0.10092628328830568, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.10092628328830568, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.10628135854882285, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.10628135854882285, lift=1.0)])]

In [34]:
frequentItemsets = list(apriori(transactions,min_support = 0.035))
frequentItemsets

[RelationRecord(items=frozenset({'20712'}), support=0.041634504052489384, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.041634504052489384, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.040283674257043614, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.040283674257043614, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04988421458896179, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04988421458896179, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.07550173678116558, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.07550173678116558, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.04911231184870706, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozense

In [35]:
frequentItemsets[-1]

RelationRecord(items=frozenset({'22699', '22697'}), support=0.037051331532226936, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.037051331532226936, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7566502463054187, lift=14.71280141221268), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7204502814258911, lift=14.71280141221268)])

RelationRecord
(items=frozenset({'22699', '22697'}), support=0.037051331532226936, 

ordered_statistics=[
OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), 
confidence=0.037051331532226936, lift=1.0), 

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7566502463054187, lift=14.71280141221268), 

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7204502814258911, lift=14.71280141221268)])

In [36]:
data_clean[data_clean['StockCode']=='22697']

InvoiceNo StockCode  ...         Country  total
1086      536524     22697  ...  United Kingdom  17.70
1644      536544     22697  ...  United Kingdom  35.46
2793      536592     22697  ...  United Kingdom  17.73
3357      536623     22697  ...  United Kingdom  35.40
3422      536627     22697  ...  United Kingdom  35.40
...          ...       ...  ...             ...    ...
539148    581433     22697  ...            EIRE  17.70
540215    581472     22697  ...  United Kingdom  35.40
540985    581492     22697  ...  United Kingdom   5.79
541220    581495     22697  ...  United Kingdom  35.40
541449    581498     22697  ...  United Kingdom  34.74

[1043 rows x 9 columns]

In [37]:
data_clean[data_clean['StockCode']=='22699']

InvoiceNo StockCode  ...         Country  total
1077      536523     22699  ...  United Kingdom  17.70
1193      536530     22699  ...  United Kingdom  11.80
1645      536544     22699  ...  United Kingdom  11.82
2794      536592     22699  ...  United Kingdom  23.64
3397      536624     22699  ...  United Kingdom  17.70
...          ...       ...  ...             ...    ...
539671    581439     22699  ...  United Kingdom  34.74
540213    581472     22699  ...  United Kingdom  35.40
540986    581492     22699  ...  United Kingdom  11.58
541218    581495     22699  ...  United Kingdom  35.40
541451    581498     22699  ...  United Kingdom  34.74

[1084 rows x 9 columns]

## (Quiz7) หา k-itemset ที่มีความน่าสนใจ (โดยพิจารณาลูกค้าเป็นรายคน) พร้อมทั้งอธิบายว่าน่าสนใจยังไง ส่งก่อน 11.25